students/Gentner/1_4GentnerButtonfinal.ino/1_4GentnerButtonfinal.ino.ino

```diff
@@ -0,0 +1,160 @@
+const int BLUE=9; // pin 9 is the blue led
+const int RED=10;// pin 10 is the red led
+const int GREEN=11; // pin 11 is the green led
+const int BUTTON=2; // setting the button as an input on pin 2
+boolean lastButton = LOW; // variable containing the previous button state
+boolean currentButton = LOW; // variable constaing the current button state
+unsigned int ledMode = 0; // present state of the LED
+unsigned long previousMillis = 0;
+unsigned int brightness = 0;
+
+void setup() {
+pinMode(9,OUTPUT);
+pinMode(10, OUTPUT);
+pinMode(11, OUTPUT); //set all pins as an output
+pinMode(BUTTON, INPUT); //set the button as an input
+}
+
+boolean debounce(boolean last)
+{
+ boolean current = digitalRead(BUTTON); // Read the button state
+ if (last != current) // if it's different...
+ {
+ delay(5); //Wait 5ms
+ current = digitalRead(BUTTON); //Read it again
+ }
+ return current; //Return the current value
+}
+
+void loop() 
+{
+  unsigned long currentMillis = millis();
+currentButton = debounce(lastButton); //Read debounced state
+ if (lastButton == LOW && currentButton == HIGH) //if it was pressed...
+ {
+ ledMode++  ; //Increase the LED state value
+ }
+ lastButton = currentButton; //Reset button value
+ if (ledMode == 6) ledMode=1;
+{
+switch (ledMode){ //tells the mode to switch
+case 1: // led off
+if (ledMode== 1)
+{
+digitalWrite(RED, LOW);
+ digitalWrite(GREEN, LOW);
+ digitalWrite(BLUE, LOW);
+ break;
+}
+else {
+break;
+}
+case 2:
+if (ledMode==2)
+{
+if (currentMillis - previousMillis >= 500)
+{previousMillis = currentMillis;
+if (digitalRead(RED)==LOW)
+{
+digitalWrite(RED, HIGH);
+ digitalWrite(GREEN, HIGH);
+ digitalWrite(BLUE, HIGH);
+}
+else {
+ digitalWrite(RED, LOW);
+ digitalWrite(GREEN, LOW);
+ digitalWrite(BLUE, LOW);
+}
+ break;
+}}
+ else {
+  break;
+ }
+
+ case 3:
+if (ledMode==3)
+{
+if (currentMillis - previousMillis >= 10)
+{previousMillis = currentMillis;
+if (brightness>=0)
+{
+for (brightness=0; brightness<256; brightness++ )
+{
+analogWrite(RED, brightness);
+ digitalWrite(GREEN, LOW);
+ analogWrite(BLUE, brightness);
+}}
+else
+for (brightness=255; brightness>=0; brightness--)
+{
+ analogWrite(RED, brightness );
+ digitalWrite(GREEN, LOW);
+ analogWrite(BLUE, brightness);
+ }
+ break;
+}}
+else {
+  break;
+ }
+
+case 4:
+if (ledMode==4)
+{
+for (int i=255; i>=0; i-- )
+{
+if (currentMillis - previousMillis >= 10)
+{previousMillis = currentMillis;
+ analogWrite(RED, i);
+ digitalWrite(GREEN, LOW);
+ analogWrite(BLUE, 0);
+}
```

Comment:
Again the logic is a little wonky. Try stepping through the code line-by-line to predict what would happen. In this case:
- `i` is set to 255
- Checks if 10ms has gone by (let's say it has in this case)
- resets timing
- sets red LED to `i`, others off. (I suspect you meant to do something like `255-i` for blue)
- returns to top of `for` loop, decreases `i` by one, so `i` is now 254
- Checks if 10ms has gone by, but it hasn't because we haven't updated `currentMillis`, so breaks the `for` loop and moves to the next `for` loop starting at `i=0`.